### PR TITLE
Multiple fixes and feature fixes for lua side of CC.

### PR DESCRIPTION
**Pull request by Wojbie**
In no particular order:

1. bios.lua
 - Added missing test for ensure turtle folder exists.
 - Fix for http.request _post type check. Was table instead of string.
4. textutils.lua
 - Added exception to complete function - will not complete if provided with LuaKeyword - solves do->dofile problem and removes other LuaKeyword related ones in advance.
 - Added support for colon operator in complete function. Will correctly autocomplete last part if colon syntax was used.
 - Removed x,y nil arguments from getCursorPos() in slowWrite().
5. edit.lua
 - Changed logic in handling the paste event - if paste event is received when menu is open it automatically closed said menu and lets it paste - resolves ctrl+ctrl+v annoyance.
- Support for colon syntax
6. set.lua
 - Switched set to use pagedPrint - this will ensure that even if there are more settings than lines on screen the set command will show you all of them.
7. startup.lua
 - Added autocompletition for turtle programs go, turn, equip, unequip. For command autocompletition please look at #309
8. lua.lua
 - Changed return function to print returned stuff correctly - will print all returned variables even if there are any nils in there.
- Support for colon syntax
9. rednet.lua
 - fix in timeout logic.
- Whitespace Cleanup
10. gps.lua
 - removed obsolete piece of code that was never running.
11. chat.lua
 - fixed up patterns used by chat program.
12. settings.lua
 - makes getNames() return sorted list - easier for eyes and same program settings are next to each other.
13. shell.lua
 - added package.config to better support require related stuff.

Dumping all those fixes/changes into one PR to simplify discussion.
This is supposed to be mixed collection of small changes that don't deserve separate PR for themselves.
Any and all changes will be fixed/changed/removed on decisions from @dan200 .

### CODE REVIEW

**Comment by SquidDev** [comment]
It might be nice if Lua printed all return values when the first one is `nil`. So replacing [this](https://github.com/dan200/ComputerCraft/blob/master/src/main/resources/assets/computercraft/lua/rom/programs/lua.lua#L67-L71) with something like:

```lua
local tResults = { pcall( func ) }
if tResults[1] then
    if tResults.n >= 2 tResults[2] ~= nil or nForcePrint ~= 0 then
        for n = 2, math.max(tResults.n, nForcePrint + 1) do
             local value = tResults[ n ]
```

Note: Totally untested and probably some room for improvement.

**Comment by Wojbie** [comment]
Well if you want to use table.n you have to use table.pack().
It can be simplified to
```lua
local tResults = table.pack( pcall( func ) )
if tResults[1] then
    local n = 1
    while n < tResults.n or (n <= nForcePrint) do```
Amanded that change,

**Comment by SquidDev** [comment]
Ahhh, I meant to use `table.pack` but totally forgot - sorry.

**Comment by Wilma456** [comment]
Call commands.list() directly is a little bit lagy. Look at #309 for a better solution.

**Comment by Wojbie** [comment]
@Wilma456 Weird. Did not notice any lag on my machine but ok. Removing that code.

**Comment by SquidDev** [comment]
@Wojbie I don't know if you'd be interested in implementing #219 and the enhancements to `textutils.pagedPrint` from #212 in this PR? They are both something I've been meaning to put together for a while, but this PR seems as good a time as any.

**Comment by Wojbie** [comment]
@SquidDev Sure. Do you have any if that made? If not I will take a look at that stuff later this week. Sadly good decreased tinkering time until weekend.

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua [hunk 1]
```diff
@@ -135,39 +135,65 @@ local tKeywords = {
...
+        local i = 0
+        repeat 
+            i=i+1
+            sMatch,nNextPosition = string.match( sLine, tPattern[i] , nPosition)
+        until sMatch
+        nColor = tPatternColors[i]
+        tColor[#tColor+1] = string.rep( tHex[ type(nColor) == "number" and nColor or  nColor(sMatch) ] , nNextPosition - nPosition)
+        nPosition = nNextPosition
     end
+    term.blit( sLine, table.concat( tColor ), string.rep( tHex[bgColour], #sLine ) )
```
Comment (SquidDev) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Do you need to blit here? Or can you keep the `term.write` & `term.setTextColour` combination?

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua [hunk 2]
```diff
@@ -135,39 +135,65 @@ local tKeywords = {
...
+    [ colors.purple ] = "a",
+    [ colors.blue ] = "b",
+    [ colors.brown ] = "c",
+    [ colors.green ] = "d",
+    [ colors.red ] = "e",
+    [ colors.black ] = "f",
+}
+
+local tPattern = {
+    [ 1 ] = "^(%-%-%[%[.-%]%])()",
```
Comment (SquidDev) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Is there any reason for explicitly including the keys here, instead of using the normal list syntax?

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua (hunk 1 above)
Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Well it blits the whole sLine at once. All Color changes are inside tColor table. This replaces multiple term.write calls with one term.blit call per line.

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua (hunk 2 above)
Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: tPatternColors contains colors that are related to this table colors. I explicitly included keys here and there to make clear which pattern in tPattern table relates to which color inside tPatternColors table. Its just to make reading this code simpler.

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua (hunk 1 above)
Comment (SquidDev) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: I know. I'm just wondering if there is any performance gain of 3 concatenations + blit vs multiple calls to `term.write` and `term.set(Text|Background)Colour`? IMO the concatenation is a little more ugly, but if it's more efficient then I'm not going to grumble :).

Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: If you look at old code it runs string.sub on sLine every time color changes. I do believe this is faster then old method.

File: src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua [hunk 3]
```diff
@@ -16,37 +23,60 @@ function loadImage( sPath )
...
     if fs.exists( sPath ) then
-        local file = io.open(sPath, "r" )
-        local sLine = file:read()
-        while sLine do
+        for sLine in io.lines(sPath) do
             local tLine = {}
-            for x=1,sLine:len() do
-                tLine[x] = tColourLookup[ string.byte(sLine,x,x) ] or 0
+            for x=1,#sLine do
+                tLine[x] = tColourLookup[ string.sub(sLine,x,x) ] or 0
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: Why this change? string.byte is a cheaper op than string.sub

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua [hunk 4]
```diff
@@ -59,6 +59,7 @@ end
 if peripheral.find( "printer" ) then
     table.insert( tMenuItems, "Print" )
 end
+table.insert( tMenuItems, "Jump" )
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Not a fan of adding another item to the menu. It complicates the UI of one of the earliest things new CC users see. The beginner flow is very important

Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Ok then. Would you say remove it then or just make a setting for it that is by default false? Like `edit.advanced_functionality` ?

File: src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua (hunk 3 above)
Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: Sure will fix that in next commit then. Any more thinks you want to change @dan200?

Comment (BombBloke) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: sub is used here so that char doesn't need to be used when rebuilding strings for paintutils.save and paintutils.drawImage.

File: src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua [hunk 5]
```diff
@@ -12,8 +12,8 @@ end
 
 local tColourLookup = {}
 for n=1,16 do
-    tColourLookup[ string.sub( "0123456789abcdef",n,n ) ] = 2^(n-1)
-    tColourLookup[ 2^(n-1) ] = string.sub( "0123456789abcdef",n,n )
+    tColourLookup[ string.byte( "0123456789abcdef",n,n ) ] = 2^(n-1)
+    tColourLookup[ 2^(n-1) ] = string.byte( "0123456789abcdef",n,n )
```
Comment (BombBloke) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: If you want to do this, then later in the code you'll also need to change some "table.concat"s into "string.char( unpack () )"s. I'm fairly sure it's faster just to use string.sub().

Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: Ohh didn't notice that... i sure dun gaffed. Should be fixed now.

**Comment by Wojbie** [comment]
Ok i am calling it stop on this PR. No more additions. All stuff i ran into that i wanted to fix on lua side is in here or in other PRs i believe this ready for review.

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua (hunk 4 above)
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Remove

File: src/main/resources/assets/computercraft/lua/bios.lua [hunk 6]
```diff
@@ -828,7 +828,7 @@ for n,sFile in ipairs( tApis ) do
     end
 end
 
-if turtle then
+if turtle and fs.isDir( "rom/apis/turtle" ) then
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/bios.lua]: Not necessary, surely? Why would turtle API exist without the mount?

File: src/main/resources/assets/computercraft/lua/rom/startup.lua [hunk 7]
```diff
@@ -205,6 +205,37 @@ shell.setCompletionFunction( "rom/programs/http/pastebin.lua", completePastebin
 shell.setCompletionFunction( "rom/programs/rednet/chat.lua", completeChat )
 shell.setCompletionFunction( "rom/programs/command/exec.lua", completeExec )
 
+if turtle then
+    local tGoOptions = { "left", "right", "forward", "back", "down", "up" }
+    local function completeGo( shell, nIndex, sText )
+        if nIndex == 1 then
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/startup.lua]: No need for this check "go left right forward up up up down right" is valid

File: src/main/resources/assets/computercraft/lua/rom/apis/rednet.lua [hunk 8]
```diff
@@ -233,7 +233,7 @@ function run()
     			if type( tMessage ) == "table" and tMessage.nMessageID then
 	    			if not tReceivedMessages[ tMessage.nMessageID ] then
 		    			tReceivedMessages[ tMessage.nMessageID ] = true
-                        tReceivedMessageTimeouts[ os.startTimer( 30 ) ] = nMessageID
+                        tReceivedMessageTimeouts[ os.startTimer( 30 ) ] = tMessage.nMessageID
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/apis/rednet.lua]: Fix whitespace while you're at it?

File: src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua [hunk 9]
```diff
@@ -189,12 +222,18 @@ function drawImage( tImage, xPos, yPos )
     if type( xPos ) ~= "number" then error( "bad argument #2 (expected number, got " .. type( xPos ) .. ")", 2 ) end
     if type( yPos ) ~= "number" then error( "bad argument #3 (expected number, got " .. type( yPos ) .. ")", 2 ) end
     for y=1,#tImage do
-        local tLine = tImage[y]
-        for x=1,#tLine do
-            if tLine[x] > 0 then
-                term.setBackgroundColor( tLine[x] )
-                drawPixelInternal( x + xPos - 1, y + yPos - 1 )
+        local tLine, sBG, counter = tImage[y], {}, 0
+        if tLine then for x=1,#tLine+1 do
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: ugly formatting

File: src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua [hunk 10]
```diff
@@ -189,12 +222,18 @@ function drawImage( tImage, xPos, yPos )
...
+                counter = counter + 1
+                sBG[counter] = tColourReverseLookup[ px ]
+            elseif counter > 0 then
+                setPos( x + xPos - 1 - counter, y + yPos - 1 )	
+                local sT = rep( " ", counter )
+                blit( sT, sT, concat( sBG ) )
+                sBG, counter = {}, 0
             end
-        end
+        end end
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/apis/paintutils.lua]: ditto

File: src/main/resources/assets/computercraft/lua/rom/help/edit.txt [hunk 11]
```diff
@@ -1,4 +1,6 @@
 edit is a text editor for creating or modifying programs or text files. After creating a program with edit, type its filename in the shell to run it. You can open any of the builtin programs with edit to learn how to program.
 
 ex:
-"edit hello" opens a file called "hello" for editing.
\ No newline at end of file
+"edit hello" opens a file called "hello" for editing.
+
+By changing setting "edit.advanced_functionality" to true you can unlock ability to jump to specific line in longer files.
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/help/edit.txt]: Remove

File: src/main/resources/assets/computercraft/lua/rom/programs/edit.lua [hunk 12]
```diff
@@ -59,6 +59,9 @@ end
 if peripheral.find( "printer" ) then
     table.insert( tMenuItems, "Print" )
 end
+if settings.get( "edit.advanced_functionality" ) then
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/rom/programs/edit.lua]: Remove

File: src/main/resources/assets/computercraft/lua/bios.lua [hunk 13]
```diff
@@ -900,6 +900,7 @@ settings.set( "shell.allow_disk_startup", (commands == nil) )
 settings.set( "shell.autocomplete", true )
 settings.set( "edit.autocomplete", true ) 
 settings.set( "edit.default_extension", "lua" )
+settings.set( "edit.advanced_functionality", false )
```
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/bios.lua]: Remove

File: src/main/resources/assets/computercraft/lua/bios.lua (hunk 6 above)
Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/bios.lua]: Well there were cheacks for `if pocket and fs.isDir( "rom/apis/pocket" ) then` and 'if commands and fs.isDir( "rom/apis/command" ) then'. I was under belief that this is accidentally committed.

Comment (Wojbie) [inline src/main/resources/assets/computercraft/lua/bios.lua]: It also makes sense to check for that on off chance that this folder accidentally (on on purpose) was removed in jar file.

**Comment by Wojbie** [comment]
I think that is all you wanted changed @dan200? 
Reverted all stuff you didn't like.

File: src/main/resources/assets/computercraft/lua/bios.lua (hunk 6 above)
Comment (dan200) [inline src/main/resources/assets/computercraft/lua/bios.lua]: Fair enough. Consistency is good,
